java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,179 @@
+      is Scope -> this + Paths(other)
+      else -> this + other
+    }
+
+    /** Returns the first [Path] in the collection, or an empty [Path]. */
+    override fun getPath(): Path = if (paths.isNotEmpty()) paths.first().getPath() else Path()
+
+    /** Substitute all aliases as a new [Paths] object. */
+    override fun substitute(aliases: Scope) =
+      Paths(paths.map { path -> path.substitute(aliases) })
+  }
+
+  /** Associates an [Identifier] with a group of [Deduction]s. */
+  data class Scope(
+    val associations: Map<Identifier, Deduction> = emptyMap()
+  ) : Deduction() {
+
+    constructor(vararg pairs: Pair<Identifier, Deduction>) : this(pairs.toMap())
+
+    /** Returns a new [Scope] with only the queried association. */
+    fun lookup(key: Identifier): Scope {
+      require(key in associations) {
+        "Scope does not associate anything with '$key'."
+      }
+      return Scope(key to associations.getOrDefault(key, Paths()))
+    }
+
+    /** Union of a [Scope] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> Paths(this) + Paths(other)
+      is Paths -> Paths(this) + other
+      is Scope -> Scope(this.associations + other.associations)
+      else -> this + other
+    }
+
+    /** [Path]s are not well defined on [Scope]s. */
+    override fun getPath() =
+      throw UnsupportedOperationException("Path not well defined on Scope object.")
+
+    /** Substitute all Aliases in each associated [Deduction] object as a new [Scope]. */
+    override fun substitute(aliases: Scope) =
+      Scope(
+        associations = associations
+          .mapKeys { (key, _) -> aliases.associations
+            .getOrDefault(key, Path(key)).getPath().path.first()
+          }
+          .mapValues { (_, Deduction) -> Deduction.substitute(aliases) }
+      )
+  }
+
+  /** Used to indicate an Equality Claim. */
+  data class Equal(val op: Deduction) : Deduction(), Pathlike {
+
+    override val path = getPath().path
+
+    /** Unwraps claim and returns underlying [Path]. */
+    override fun getPath(): Path = op.getPath()
+
+    /** Apply alias substitutions for [Equal] claim. */
+    override fun substitute(aliases: Scope) = Equal(op.substitute(aliases))
+
+    /** Union of an [Equal] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(op + other.op)
+      is Derive -> Derive(op + other.op)
+      else -> Derive(op + other)
+    }
+
+    /** Append [Identifier]s to wrapped [Path]. */
+    override fun mergeTop(other: Pathlike): Deduction = when (op) {
+      is Pathlike -> Equal(op.mergeTop(other))
+      else -> throw UnsupportedOperationException(
+        "Contained Deduction ${op::class.simpleName} does not permit `mergeTop` operation."
+      )
+    }
+  }
+
+  /** Used to indicate a Derivation Claim. */
+  data class Derive(val op: Deduction) : Deduction(), Pathlike {
```

Comment:
Derive doesn't have a single path necessarily, how can it be Pathlike?

Reply:
That's a good catch. I removed the interface, and it worked fine :)